cmd/maestro/server/pulse_server.go

```diff
@@ -0,0 +1,114 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/openshift-online/maestro/pkg/api"
+	"github.com/openshift-online/maestro/pkg/client/cloudevents"
+	"github.com/openshift-online/maestro/pkg/dao"
+	"github.com/openshift-online/maestro/pkg/db"
+	"github.com/openshift-online/maestro/pkg/logger"
+)
+
+// PulseServer is a server that periodically updates the heartbeat for the current Maestro instance,
+// checks all Maestro instances' liveness and trigger statusresync for dead instances.
+type PulseServer struct {
+	instanceID    string
+	pulseInterval int64
+	checkInterval int64
+	instanceDao   dao.InstanceDao
+	lockFactory   db.LockFactory
+	sourceClient  cloudevents.SourceClient
+}
+
+func NewPulseServer() *PulseServer {
+	sessionFactory := env().Database.SessionFactory
+	return &PulseServer{
+		instanceID:    env().Config.MessageBroker.ClientID,
+		pulseInterval: env().Config.PulseServer.PulseInterval,
+		checkInterval: env().Config.PulseServer.CheckInterval,
+		instanceDao:   dao.NewInstanceDao(&sessionFactory),
+		lockFactory:   db.NewAdvisoryLockFactory(sessionFactory),
+		sourceClient:  env().Clients.CloudEventsSource,
+	}
+}
+
+// Start initializes and runs the pulse server.
+// It periodically updates the heartbeat for the current Maestro instance,
+// checks Maestro instances' liveness and trigger statusresync for dead instances.
+func (s *PulseServer) Start(ctx context.Context) {
+	logger := logger.NewOCMLogger(ctx)
+	instance := &api.ServerInstance{
+		Meta: api.Meta{
+			ID:        s.instanceID,
+			UpdatedAt: time.Now(),
+		},
+	}
+	_, err := s.instanceDao.UpSert(ctx, instance)
+	if err != nil {
+		logger.Error(fmt.Sprintf("Unable to create maestro instance: %s", err.Error()))
+		return
+	}
+
+	pulseTicker := time.NewTicker(time.Duration(s.pulseInterval) * time.Second)
+	checkTicker := time.NewTicker(time.Duration(s.checkInterval) * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			pulseTicker.Stop()
+			checkTicker.Stop()
+			return
+		case <-pulseTicker.C:
+			logger.Infof("Updating heartbeat for maestro instance: %s", instance.ID)
+			instance.UpdatedAt = time.Now()
+			_, err = s.instanceDao.UpSert(ctx, instance)
+			if err != nil {
+				// ignore the error and continue to tolerate the intermittent db connection issue
+				logger.Error(fmt.Sprintf("Unable to update heartbeat for maestro instance: %s", err.Error()))
+				continue
+			}
+		case <-checkTicker.C:
+			logger.Infof("Checking maestro instances liveness and trigger statusresync for dead instances")
+			// lock the Instance with a fail-fast advisory lock context.
+			// this allows concurrent processing of many instances by one or more maestro instances exclusively.
+			lockOwnerID, acquired, err := s.lockFactory.NewNonBlockingLock(ctx, "maestro-instances-pulse-check", db.Instances)
+			if err != nil {
+				logger.Error(fmt.Sprintf("Unable to obtain the event lock: %s", err.Error()))
```

Comment:
also need Unlock here, because we already start a transaction here, we need ensure to end it

Reply:
and cloud we consider use `defer` for unlock

Reply:
`defer` doesn't suit for this case, right? we want to unlock when the current loop is done.

Reply:
yeah, I saw the unlock at the end, but when you call `continue`, it will not be executed, right?

Reply:
for `defer`, could we have a function for this, I think it will help us to ensure the unlock is called

Reply:
we also need this before https://github.com/openshift-online/maestro/pull/36/files#diff-088f089b05000e4281ad5507d12cae646fe7245b34eb843047d70244ef516e3aR79 and https://github.com/openshift-online/maestro/pull/36/files#diff-088f089b05000e4281ad5507d12cae646fe7245b34eb843047d70244ef516e3aR84

and I think it's better to create a new function for this part and use `defer` in that function to avoid unlock :-)

Reply:
do we need to unlock if the lock is not acquired?

Reply:
yes, when you call new lock, we will start a db transaction at first, you need to call unlock to end the db transaction

Reply:
done.